src/persist/src/indexed/mod.rs

```diff
@@ -790,82 +789,57 @@ impl<L: Log, B: Blob> Indexed<L, B> {
             Err(_) => {
-                // TODO: we should ignore this error if it doesn't correspond
-                // to the currently in-flight request, to avoid sending the
-                // Maintainer duplicate requests if it sends us duplicated/stale
-                // errors.
                 self.metrics.trace_compaction_error_response_count.inc();
                 return;
             }
         };
 
-        // Check that the returned response is indeed for the matching in-flight request.
-        // We expect this condition to always be true while the connection between Indexed
-        // and Maintainer is infallible.
-        soft_assert_or_log!(sent_compact_trace_req == res.req,
-                "received invalid trace compaction response for stream {:?} currently in flight req {:?}: {:?}. Ignoring",
-                id, sent_compact_trace_req, res
-            );
-
-        // Reinstate the in-flight request if the response doesn't match, and ignore the response.
-        //
-        // TODO: can we simply handle the response? This is complicated for two reasons.
-        //
-        // 1. Using the results of a request that is not the same request we sent risks
-        //    breaking the invariant that trace batch compaction levels are nonincreasing.
-        // 2. We only want to change the "has a request in flight" status of the persisted
-        //    stream if the non-matching response is actually useful, (as otherwise we'd
-        //    rather wait for the matching response before changing the status). We don't
-        //    know if the response is useful in advance, but we don't have access to the
-        //    `in_flight_trace_compactions` within `apply_batched_cmd`. The second bullet
-        //    is more manageable than the first e.g. by returning a boolean from the
-        //    closure passed to `apply_batched_cmd`.
-
-        if sent_compact_trace_req != res.req {
-            // Restore our bookkeeping of which request is in flight.
-            self.in_flight_trace_compactions
-                .insert(id, sent_compact_trace_req);
-            return;
-        }
-
-        let written_bytes = res.merged.size_bytes;
+        let written_bytes = response.merged.size_bytes;
         self.metrics.compaction_write_bytes.inc_by(written_bytes);
 
         self.apply_batched_cmd(|state, pending| {
-            let arrangement = state
-                .arrangements
-                .get_mut(&id)
-                .expect("arrangement known to exist");
-            let (_, deleted_trace_batches) = arrangement.trace_handle_compact_response(res);
+            // Ignore the response if it belongs to a stream that has since been
+            // deleted.
+            let arrangement = match state.arrangements.get_mut(&id) {
+                Some(arrangement) => arrangement,
+                None => return,
+            };
+            let (_, deleted_trace_batches) = arrangement.trace_handle_compact_response(response);
```

Comment:
right. this is a holdover from the previous api where `trace_step` did the compaction and returned the number of bytes it wrote. we actually already increment the relevant metric a few lines up

in the most recent version of the code it looks like:

```
1057         let written_bytes = response.merged.size_bytes;
1058         self.metrics.compaction_write_bytes.inc_by(written_bytes);
1059
1060         self.apply_batched_cmd(|state, pending| {
1061             // Ignore the response if it belongs to a stream that has since been
1062             // deleted.
1063             let arrangement = match state.arrangements.get_mut(&id) {
1064                 Some(arrangement) => arrangement,
1065                 None => return,
1066             };
1067             let deleted_trace_batches = arrangement.trace_handle_compact_response(response);
1068             if !deleted_trace_batches.is_empty() {
1069                 pending.add_deleted_trace_batches(deleted_trace_batches);
1070             }
1071         });
```